// ng-http-loader/bundles/ng-http-loader.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common/http'), require('rxjs'), require('rxjs/operators'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-http-loader', ['exports', '@angular/core', '@angular/common/http', 'rxjs', 'rxjs/operators', '@angular/common'], factory) :
    (factory((global['ng-http-loader'] = {}),global.ng.core,global.ng.common.http,global.rxjs,global.rxjs.operators,global.ng.common));
}(this, (function (exports,i0,http,rxjs,operators,common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /**
     * @abstract
     */
    var AbstractLoader = /** @class */ (function () {
        function AbstractLoader() {
        }
        AbstractLoader.propDecorators = {
            backgroundColor: [{ type: i0.Input }]
        };
        return AbstractLoader;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkChasingDotsComponent = /** @class */ (function (_super) {
        __extends(SkChasingDotsComponent, _super);
        function SkChasingDotsComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkChasingDotsComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-chasing-dots',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-chasing-dots\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-child sk-dot1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-dot2\" [style.background-color]='backgroundColor'></div>\n</div>\n",
                        styles: [".sk-chasing-dots{top:50%;margin:auto;width:40px;height:40px;position:relative;text-align:center;-webkit-animation:2s linear infinite sk-chasingDotsRotate;animation:2s linear infinite sk-chasingDotsRotate}.sk-chasing-dots .sk-child{width:60%;height:60%;display:inline-block;position:absolute;top:0;border-radius:100%;-webkit-animation:2s ease-in-out infinite sk-chasingDotsBounce;animation:2s ease-in-out infinite sk-chasingDotsBounce}.sk-chasing-dots .sk-dot2{top:auto;bottom:0;-webkit-animation-delay:-1s;animation-delay:-1s}@-webkit-keyframes sk-chasingDotsRotate{100%{-webkit-transform:rotate(360deg);transform:rotate(360deg)}}@keyframes sk-chasingDotsRotate{100%{-webkit-transform:rotate(360deg);transform:rotate(360deg)}}@-webkit-keyframes sk-chasingDotsBounce{0%,100%{-webkit-transform:scale(0);transform:scale(0)}50%{-webkit-transform:scale(1);transform:scale(1)}}@keyframes sk-chasingDotsBounce{0%,100%{-webkit-transform:scale(0);transform:scale(0)}50%{-webkit-transform:scale(1);transform:scale(1)}}"]
                    }] }
        ];
        return SkChasingDotsComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkCubeGridComponent = /** @class */ (function (_super) {
        __extends(SkCubeGridComponent, _super);
        function SkCubeGridComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkCubeGridComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-cube-grid',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-cube-grid\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-cube sk-cube1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube2\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube3\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube4\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube5\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube6\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube7\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube8\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube9\" [style.background-color]='backgroundColor'></div>\n</div>\n",
                        styles: [".sk-cube-grid{position:relative;top:50%;width:40px;height:40px;margin:auto}.sk-cube-grid .sk-cube{width:33.33%;height:33.33%;float:left;-webkit-animation:1.3s ease-in-out infinite sk-cubeGridScaleDelay;animation:1.3s ease-in-out infinite sk-cubeGridScaleDelay}.sk-cube-grid .sk-cube1{-webkit-animation-delay:.2s;animation-delay:.2s}.sk-cube-grid .sk-cube2{-webkit-animation-delay:.3s;animation-delay:.3s}.sk-cube-grid .sk-cube3{-webkit-animation-delay:.4s;animation-delay:.4s}.sk-cube-grid .sk-cube4{-webkit-animation-delay:.1s;animation-delay:.1s}.sk-cube-grid .sk-cube5{-webkit-animation-delay:.2s;animation-delay:.2s}.sk-cube-grid .sk-cube6{-webkit-animation-delay:.3s;animation-delay:.3s}.sk-cube-grid .sk-cube7{-webkit-animation-delay:0ms;animation-delay:0ms}.sk-cube-grid .sk-cube8{-webkit-animation-delay:.1s;animation-delay:.1s}.sk-cube-grid .sk-cube9{-webkit-animation-delay:.2s;animation-delay:.2s}@-webkit-keyframes sk-cubeGridScaleDelay{0%,100%,70%{-webkit-transform:scale3D(1,1,1);transform:scale3D(1,1,1)}35%{-webkit-transform:scale3D(0,0,1);transform:scale3D(0,0,1)}}@keyframes sk-cubeGridScaleDelay{0%,100%,70%{-webkit-transform:scale3D(1,1,1);transform:scale3D(1,1,1)}35%{-webkit-transform:scale3D(0,0,1);transform:scale3D(0,0,1)}}"]
                    }] }
        ];
        return SkCubeGridComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkDoubleBounceComponent = /** @class */ (function (_super) {
        __extends(SkDoubleBounceComponent, _super);
        function SkDoubleBounceComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkDoubleBounceComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-double-bounce',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-double-bounce\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-child sk-double-bounce1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-double-bounce2\" [style.background-color]='backgroundColor'></div>\n</div>\n",
                        styles: [".sk-double-bounce{top:50%;width:40px;height:40px;position:relative;margin:auto}.sk-double-bounce .sk-child{width:100%;height:100%;border-radius:50%;background-color:#333;opacity:.6;position:absolute;top:0;left:0;-webkit-animation:2s ease-in-out infinite sk-doubleBounce;animation:2s ease-in-out infinite sk-doubleBounce}.sk-double-bounce .sk-double-bounce2{-webkit-animation-delay:-1s;animation-delay:-1s}@-webkit-keyframes sk-doubleBounce{0%,100%{-webkit-transform:scale(0);transform:scale(0)}50%{-webkit-transform:scale(1);transform:scale(1)}}@keyframes sk-doubleBounce{0%,100%{-webkit-transform:scale(0);transform:scale(0)}50%{-webkit-transform:scale(1);transform:scale(1)}}"]
                    }] }
        ];
        return SkDoubleBounceComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkRotatingPlaneComponent = /** @class */ (function (_super) {
        __extends(SkRotatingPlaneComponent, _super);
        function SkRotatingPlaneComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkRotatingPlaneComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-rotating-plane',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-rotating-plane colored-parent\" [style.background-color]='backgroundColor'></div>\n",
                        styles: [".sk-rotating-plane{position:relative;top:50%;width:40px;height:40px;margin:auto;-webkit-animation:1.2s ease-in-out infinite sk-rotatePlane;animation:1.2s ease-in-out infinite sk-rotatePlane}@-webkit-keyframes sk-rotatePlane{0%{-webkit-transform:perspective(120px) rotateX(0) rotateY(0);transform:perspective(120px) rotateX(0) rotateY(0)}50%{-webkit-transform:perspective(120px) rotateX(-180.1deg) rotateY(0);transform:perspective(120px) rotateX(-180.1deg) rotateY(0)}100%{-webkit-transform:perspective(120px) rotateX(-180deg) rotateY(-179.9deg);transform:perspective(120px) rotateX(-180deg) rotateY(-179.9deg)}}@keyframes sk-rotatePlane{0%{-webkit-transform:perspective(120px) rotateX(0) rotateY(0);transform:perspective(120px) rotateX(0) rotateY(0)}50%{-webkit-transform:perspective(120px) rotateX(-180.1deg) rotateY(0);transform:perspective(120px) rotateX(-180.1deg) rotateY(0)}100%{-webkit-transform:perspective(120px) rotateX(-180deg) rotateY(-179.9deg);transform:perspective(120px) rotateX(-180deg) rotateY(-179.9deg)}}"]
                    }] }
        ];
        return SkRotatingPlaneComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkSpinnerPulseComponent = /** @class */ (function (_super) {
        __extends(SkSpinnerPulseComponent, _super);
        function SkSpinnerPulseComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkSpinnerPulseComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-spinner-pulse',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-spinner sk-spinner-pulse colored-parent\" [style.background-color]='backgroundColor'></div>\n",
                        styles: [".sk-spinner-pulse{position:relative;top:50%;width:40px;height:40px;margin:auto;border-radius:100%;-webkit-animation:1s ease-in-out infinite sk-pulseScaleOut;animation:1s ease-in-out infinite sk-pulseScaleOut}@-webkit-keyframes sk-pulseScaleOut{0%{-webkit-transform:scale(0);transform:scale(0)}100%{-webkit-transform:scale(1);transform:scale(1);opacity:0}}@keyframes sk-pulseScaleOut{0%{-webkit-transform:scale(0);transform:scale(0)}100%{-webkit-transform:scale(1);transform:scale(1);opacity:0}}"]
                    }] }
        ];
        return SkSpinnerPulseComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkThreeBounceComponent = /** @class */ (function (_super) {
        __extends(SkThreeBounceComponent, _super);
        function SkThreeBounceComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkThreeBounceComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-three-bounce',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-three-bounce\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-child sk-bounce1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-bounce2\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-child sk-bounce3\" [style.background-color]='backgroundColor'></div>\n</div>\n",
                        styles: [".sk-three-bounce{top:50%;position:relative;margin:auto;width:80px;text-align:center}.sk-three-bounce .sk-child{width:20px;height:20px;border-radius:100%;display:inline-block;-webkit-animation:1.4s ease-in-out infinite both sk-three-bounce;animation:1.4s ease-in-out infinite both sk-three-bounce}.sk-three-bounce .sk-bounce1{-webkit-animation-delay:-.32s;animation-delay:-.32s}.sk-three-bounce .sk-bounce2{-webkit-animation-delay:-.16s;animation-delay:-.16s}@-webkit-keyframes sk-three-bounce{0%,100%,80%{-webkit-transform:scale(0);transform:scale(0)}40%{-webkit-transform:scale(1);transform:scale(1)}}@keyframes sk-three-bounce{0%,100%,80%{-webkit-transform:scale(0);transform:scale(0)}40%{-webkit-transform:scale(1);transform:scale(1)}}"]
                    }] }
        ];
        return SkThreeBounceComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkWanderingCubesComponent = /** @class */ (function (_super) {
        __extends(SkWanderingCubesComponent, _super);
        function SkWanderingCubesComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkWanderingCubesComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-wandering-cubes',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-wandering-cubes\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-cube sk-cube1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-cube sk-cube2\" [style.background-color]='backgroundColor'></div>\n</div>\n",
                        styles: [".sk-wandering-cubes{top:50%;margin:auto;width:40px;height:40px;position:relative}.sk-wandering-cubes .sk-cube{width:10px;height:10px;position:absolute;top:0;left:0;-webkit-animation:1.8s ease-in-out -1.8s infinite both sk-wanderingCube;animation:1.8s ease-in-out -1.8s infinite both sk-wanderingCube}.sk-wandering-cubes .sk-cube2{-webkit-animation-delay:-.9s;animation-delay:-.9s}@-webkit-keyframes sk-wanderingCube{0%{-webkit-transform:rotate(0);transform:rotate(0)}25%{-webkit-transform:translateX(30px) rotate(-90deg) scale(.5);transform:translateX(30px) rotate(-90deg) scale(.5)}50%{-webkit-transform:translateX(30px) translateY(30px) rotate(-179deg);transform:translateX(30px) translateY(30px) rotate(-179deg)}50.1%{-webkit-transform:translateX(30px) translateY(30px) rotate(-180deg);transform:translateX(30px) translateY(30px) rotate(-180deg)}75%{-webkit-transform:translateX(0) translateY(30px) rotate(-270deg) scale(.5);transform:translateX(0) translateY(30px) rotate(-270deg) scale(.5)}100%{-webkit-transform:rotate(-360deg);transform:rotate(-360deg)}}@keyframes sk-wanderingCube{0%{-webkit-transform:rotate(0);transform:rotate(0)}25%{-webkit-transform:translateX(30px) rotate(-90deg) scale(.5);transform:translateX(30px) rotate(-90deg) scale(.5)}50%{-webkit-transform:translateX(30px) translateY(30px) rotate(-179deg);transform:translateX(30px) translateY(30px) rotate(-179deg)}50.1%{-webkit-transform:translateX(30px) translateY(30px) rotate(-180deg);transform:translateX(30px) translateY(30px) rotate(-180deg)}75%{-webkit-transform:translateX(0) translateY(30px) rotate(-270deg) scale(.5);transform:translateX(0) translateY(30px) rotate(-270deg) scale(.5)}100%{-webkit-transform:rotate(-360deg);transform:rotate(-360deg)}}"]
                    }] }
        ];
        return SkWanderingCubesComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SkWaveComponent = /** @class */ (function (_super) {
        __extends(SkWaveComponent, _super);
        function SkWaveComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        SkWaveComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sk-wave',
                        template: "<!--\nCopyright (c) 2015 Tobias Ahlin\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n-->\n\n<div class=\"sk-wave\" [class.colored]=\"!backgroundColor\">\n    <div class=\"sk-rect sk-rect1\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect2\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect3\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect4\" [style.background-color]='backgroundColor'></div>\n    <div class=\"sk-rect sk-rect5\" [style.background-color]='backgroundColor'></div>\n</div>\n",
                        styles: [".sk-wave{position:relative;top:50%;margin:auto;width:50px;height:40px;text-align:center;font-size:10px}.sk-wave .sk-rect{float:left;margin-right:1px;height:100%;width:6px;display:inline-block;-webkit-animation:1.2s ease-in-out infinite sk-waveStretchDelay;animation:1.2s ease-in-out infinite sk-waveStretchDelay}.sk-wave .sk-rect1{-webkit-animation-delay:-1.2s;animation-delay:-1.2s}.sk-wave .sk-rect2{-webkit-animation-delay:-1.1s;animation-delay:-1.1s}.sk-wave .sk-rect3{-webkit-animation-delay:-1s;animation-delay:-1s}.sk-wave .sk-rect4{-webkit-animation-delay:-.9s;animation-delay:-.9s}.sk-wave .sk-rect5{-webkit-animation-delay:-.8s;animation-delay:-.8s}@-webkit-keyframes sk-waveStretchDelay{0%,100%,40%{-webkit-transform:scaleY(.4);transform:scaleY(.4)}20%{-webkit-transform:scaleY(1);transform:scaleY(1)}}@keyframes sk-waveStretchDelay{0%,100%,40%{-webkit-transform:scaleY(.4);transform:scaleY(.4)}20%{-webkit-transform:scaleY(1);transform:scaleY(1)}}"]
                    }] }
        ];
        return SkWaveComponent;
    }(AbstractLoader));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var PendingInterceptorService = /** @class */ (function () {
        function PendingInterceptorService() {
            this._pendingRequests = 0;
            this._pendingRequestsStatus = new rxjs.ReplaySubject(1);
            this._filteredUrlPatterns = [];
            this._filteredMethods = [];
            this._filteredHeaders = [];
        }
        Object.defineProperty(PendingInterceptorService.prototype, "pendingRequestsStatus$", {
            get: /**
             * @return {?}
             */ function () {
                return this._pendingRequestsStatus.asObservable();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PendingInterceptorService.prototype, "pendingRequests", {
            get: /**
             * @return {?}
             */ function () {
                return this._pendingRequests;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PendingInterceptorService.prototype, "filteredUrlPatterns", {
            get: /**
             * @return {?}
             */ function () {
                return this._filteredUrlPatterns;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PendingInterceptorService.prototype, "filteredMethods", {
            set: /**
             * @param {?} httpMethods
             * @return {?}
             */ function (httpMethods) {
                this._filteredMethods = httpMethods;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PendingInterceptorService.prototype, "filteredHeaders", {
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._filteredHeaders = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PendingInterceptorService.prototype, "forceByPass", {
            set: /**
             * @param {?} value
             * @return {?}
             */ function (value) {
                this._forceByPass = value;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} url
         * @return {?}
         */
        PendingInterceptorService.prototype.shouldBypassUrl = /**
         * @param {?} url
         * @return {?}
         */
            function (url) {
                return this._filteredUrlPatterns.some(function (e) {
                    return e.test(url);
                });
            };
        /**
         * @param {?} req
         * @return {?}
         */
        PendingInterceptorService.prototype.shouldBypassMethod = /**
         * @param {?} req
         * @return {?}
         */
            function (req) {
                return this._filteredMethods.some(function (e) {
                    return e.toUpperCase() === req.method.toUpperCase();
                });
            };
        /**
         * @param {?} req
         * @return {?}
         */
        PendingInterceptorService.prototype.shouldBypassHeader = /**
         * @param {?} req
         * @return {?}
         */
            function (req) {
                return this._filteredHeaders.some(function (e) {
                    return req.headers.has(e);
                });
            };
        /**
         * @param {?} req
         * @return {?}
         */
        PendingInterceptorService.prototype.shouldBypass = /**
         * @param {?} req
         * @return {?}
         */
            function (req) {
                return this.shouldBypassUrl(req.urlWithParams)
                    || this.shouldBypassMethod(req)
                    || this.shouldBypassHeader(req)
                    || this._forceByPass;
            };
        /**
         * @param {?} req
         * @param {?} next
         * @return {?}
         */
        PendingInterceptorService.prototype.intercept = /**
         * @param {?} req
         * @param {?} next
         * @return {?}
         */
            function (req, next) {
                var _this = this;
                /** @type {?} */
                var shouldBypass = this.shouldBypass(req);
                if (!shouldBypass) {
                    this._pendingRequests++;
                    if (1 === this._pendingRequests) {
                        this._pendingRequestsStatus.next(true);
                    }
                }
                return next.handle(req).pipe(operators.map(function (event) {
                    return event;
                }), operators.catchError(function (error) {
                    return rxjs.throwError(error);
                }), operators.finalize(function () {
                    if (!shouldBypass) {
                        _this._pendingRequests--;
                        if (0 === _this._pendingRequests) {
                            _this._pendingRequestsStatus.next(false);
                        }
                    }
                }));
            };
        PendingInterceptorService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */ PendingInterceptorService.ngInjectableDef = i0.defineInjectable({ factory: function PendingInterceptorService_Factory() { return new PendingInterceptorService(); }, token: PendingInterceptorService, providedIn: "root" });
        return PendingInterceptorService;
    }());
    /** @type {?} */
    var PendingInterceptorServiceInterceptor = [{
            provide: http.HTTP_INTERCEPTORS,
            useExisting: PendingInterceptorService,
            multi: true
        }];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var SpinnerVisibilityService = /** @class */ (function () {
        function SpinnerVisibilityService(pendingInterceptorService) {
            this.pendingInterceptorService = pendingInterceptorService;
            this._visibilitySubject = new rxjs.ReplaySubject(1);
        }
        Object.defineProperty(SpinnerVisibilityService.prototype, "visibilityObservable$", {
            get: /**
             * @return {?}
             */ function () {
                return this._visibilitySubject.asObservable();
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        SpinnerVisibilityService.prototype.show = /**
         * @return {?}
         */
            function () {
                this.pendingInterceptorService.forceByPass = true;
                this._visibilitySubject.next(true);
            };
        /**
         * @return {?}
         */
        SpinnerVisibilityService.prototype.hide = /**
         * @return {?}
         */
            function () {
                this._visibilitySubject.next(false);
                this.pendingInterceptorService.forceByPass = false;
            };
        SpinnerVisibilityService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        SpinnerVisibilityService.ctorParameters = function () {
            return [
                { type: PendingInterceptorService }
            ];
        };
        /** @nocollapse */ SpinnerVisibilityService.ngInjectableDef = i0.defineInjectable({ factory: function SpinnerVisibilityService_Factory() { return new SpinnerVisibilityService(i0.inject(PendingInterceptorService)); }, token: SpinnerVisibilityService, providedIn: "root" });
        return SpinnerVisibilityService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @type {?} */
    var Spinkit = {
        skChasingDots: 'sk-chasing-dots',
        skCubeGrid: 'sk-cube-grid',
        skDoubleBounce: 'sk-double-bounce',
        skRotatingPlane: 'sk-rotationg-plane',
        skSpinnerPulse: 'sk-spinner-pulse',
        skThreeBounce: 'sk-three-bounce',
        skWanderingCubes: 'sk-wandering-cubes',
        skWave: 'sk-wave',
    };
    /** @type {?} */
    var SPINKIT_COMPONENTS = [
        SkCubeGridComponent,
        SkChasingDotsComponent,
        SkDoubleBounceComponent,
        SkRotatingPlaneComponent,
        SkSpinnerPulseComponent,
        SkThreeBounceComponent,
        SkWanderingCubesComponent,
        SkWaveComponent,
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgHttpLoaderComponent = /** @class */ (function () {
        function NgHttpLoaderComponent(pendingInterceptorService, spinnerVisibilityService) {
            var _this = this;
            this.pendingInterceptorService = pendingInterceptorService;
            this.spinnerVisibilityService = spinnerVisibilityService;
            this.spinkit = Spinkit;
            this.visibleUntil = Date.now();
            this.spinner = Spinkit.skCubeGrid;
            this.filteredUrlPatterns = [];
            this.filteredMethods = [];
            this.filteredHeaders = [];
            this.debounceDelay = 0;
            this.minDuration = 0;
            this.extraDuration = 0;
            this.entryComponent = null;
            var _a = __read(operators.partition(function (h) { return h; })(this.pendingInterceptorService.pendingRequestsStatus$), 2), showSpinner$ = _a[0], hideSpinner$ = _a[1];
            this.subscriptions = rxjs.merge(this.pendingInterceptorService.pendingRequestsStatus$.pipe(operators.switchMap(function () { return showSpinner$.pipe(operators.debounce(function () { return rxjs.timer(_this.debounceDelay); })); })), showSpinner$.pipe(operators.switchMap(function () { return hideSpinner$.pipe(operators.debounce(function () { return _this.getVisibilityTimer(); })); })), this.spinnerVisibilityService.visibilityObservable$)
                .pipe(operators.distinctUntilChanged())
                .subscribe(function (h) { return _this.handleSpinnerVisibility(h); });
        }
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.nullifySpinnerIfEntryComponentIsDefined();
                this.initFilters();
            };
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.subscriptions.unsubscribe();
            };
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.nullifySpinnerIfEntryComponentIsDefined = /**
         * @return {?}
         */
            function () {
                if (null != this.entryComponent) {
                    this.spinner = null;
                }
            };
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.initFilters = /**
         * @return {?}
         */
            function () {
                this.initFilteredUrlPatterns();
                this.initFilteredMethods();
                this.initFilteredHeaders();
            };
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.initFilteredUrlPatterns = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (!(this.filteredUrlPatterns instanceof Array)) {
                    throw new TypeError('`filteredUrlPatterns` must be an array.');
                }
                if (!!this.filteredUrlPatterns.length) {
                    this.filteredUrlPatterns.forEach(function (e) {
                        return _this.pendingInterceptorService.filteredUrlPatterns.push(new RegExp(e));
                    });
                }
            };
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.initFilteredMethods = /**
         * @return {?}
         */
            function () {
                if (!(this.filteredMethods instanceof Array)) {
                    throw new TypeError('`filteredMethods` must be an array.');
                }
                this.pendingInterceptorService.filteredMethods = this.filteredMethods;
            };
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.initFilteredHeaders = /**
         * @return {?}
         */
            function () {
                if (!(this.filteredHeaders instanceof Array)) {
                    throw new TypeError('`filteredHeaders` must be an array.');
                }
                this.pendingInterceptorService.filteredHeaders = this.filteredHeaders;
            };
        /**
         * @param {?} showSpinner
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.handleSpinnerVisibility = /**
         * @param {?} showSpinner
         * @return {?}
         */
            function (showSpinner) {
                if (showSpinner) {
                    this.visibleUntil = Date.now() + this.minDuration;
                }
                this.isSpinnerVisible = showSpinner;
            };
        /**
         * @return {?}
         */
        NgHttpLoaderComponent.prototype.getVisibilityTimer = /**
         * @return {?}
         */
            function () {
                return rxjs.timer(Math.max(this.extraDuration, this.visibleUntil - Date.now()));
            };
        NgHttpLoaderComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'ng-http-loader',
                        template: "<div id=\"spinner\" *ngIf=\"isSpinnerVisible\">\n\n    <ng-container *ngComponentOutlet=\"entryComponent\"></ng-container>\n\n    <sk-cube-grid\n        *ngIf=\"spinner === spinkit.skCubeGrid\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-cube-grid>\n\n    <sk-chasing-dots\n        *ngIf=\"spinner === spinkit.skChasingDots\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-chasing-dots>\n\n    <sk-double-bounce\n        *ngIf=\"spinner === spinkit.skDoubleBounce\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-double-bounce>\n\n    <sk-rotating-plane\n        *ngIf=\"spinner === spinkit.skRotatingPlane\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-rotating-plane>\n\n    <sk-spinner-pulse\n        *ngIf=\"spinner === spinkit.skSpinnerPulse\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-spinner-pulse>\n\n    <sk-three-bounce\n        *ngIf=\"spinner === spinkit.skThreeBounce\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-three-bounce>\n\n    <sk-wandering-cubes\n        *ngIf=\"spinner === spinkit.skWanderingCubes\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-wandering-cubes>\n\n    <sk-wave\n        *ngIf=\"spinner === spinkit.skWave\"\n        [backgroundColor]=\"backgroundColor\">\n    </sk-wave>\n\n</div>\n\n",
                        styles: ["#spinner{top:0;left:0;height:100%;width:100%;position:fixed;z-index:9999;display:flex;align-items:center;justify-content:center;opacity:.7;background-color:#f1f1f1}::ng-deep .colored-parent,::ng-deep .colored>div{background-color:#333}"]
                    }] }
        ];
        /** @nocollapse */
        NgHttpLoaderComponent.ctorParameters = function () {
            return [
                { type: PendingInterceptorService },
                { type: SpinnerVisibilityService }
            ];
        };
        NgHttpLoaderComponent.propDecorators = {
            backgroundColor: [{ type: i0.Input }],
            spinner: [{ type: i0.Input }],
            filteredUrlPatterns: [{ type: i0.Input }],
            filteredMethods: [{ type: i0.Input }],
            filteredHeaders: [{ type: i0.Input }],
            debounceDelay: [{ type: i0.Input }],
            minDuration: [{ type: i0.Input }],
            extraDuration: [{ type: i0.Input }],
            entryComponent: [{ type: i0.Input }]
        };
        return NgHttpLoaderComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgHttpLoaderModule = /** @class */ (function () {
        function NgHttpLoaderModule() {
        }
        NgHttpLoaderModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: __spread([
                            NgHttpLoaderComponent
                        ], SPINKIT_COMPONENTS),
                        imports: [
                            common.CommonModule,
                            http.HttpClientModule,
                        ],
                        exports: __spread([
                            NgHttpLoaderComponent
                        ], SPINKIT_COMPONENTS),
                        providers: [
                            PendingInterceptorServiceInterceptor,
                        ]
                    },] }
        ];
        return NgHttpLoaderModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.SkChasingDotsComponent = SkChasingDotsComponent;
    exports.SkCubeGridComponent = SkCubeGridComponent;
    exports.SkDoubleBounceComponent = SkDoubleBounceComponent;
    exports.SkRotatingPlaneComponent = SkRotatingPlaneComponent;
    exports.SkSpinnerPulseComponent = SkSpinnerPulseComponent;
    exports.SkThreeBounceComponent = SkThreeBounceComponent;
    exports.SkWanderingCubesComponent = SkWanderingCubesComponent;
    exports.SkWaveComponent = SkWaveComponent;
    exports.NgHttpLoaderComponent = NgHttpLoaderComponent;
    exports.AbstractLoader = AbstractLoader;
    exports.PendingInterceptorService = PendingInterceptorService;
    exports.PendingInterceptorServiceInterceptor = PendingInterceptorServiceInterceptor;
    exports.SpinnerVisibilityService = SpinnerVisibilityService;
    exports.NgHttpLoaderModule = NgHttpLoaderModule;
    exports.Spinkit = Spinkit;
    exports.SPINKIT_COMPONENTS = SPINKIT_COMPONENTS;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
